darshan-util/pydarshan/darshan/report.py

```diff
@@ -604,13 +604,13 @@ def mod_read_all_records(self, mod, dtype=None, warnings=True):
 
 
         # fetch records
-        rec = backend.log_get_generic_record(self.log, mod, dtype=dtype)
+        rec = backend.log_get_generic_record(self.log, mod, c_cols=cn, fc_cols=fcn, dtype=dtype)
         while rec != None:
             self.records[mod].append(rec)
             self._modules[mod]['num_records'] += 1
 
             # fetch next
-            rec = backend.log_get_generic_record(self.log, mod, dtype=dtype)
+            rec = backend.log_get_generic_record(self.log, mod, c_cols=cn, fc_cols=fcn, dtype=dtype)
```

Comment:
I can't find a reason to change the function signature when I check locally, using an LRU cache is apparently even faster (on top of still passing tests, and keeping the same API). The calls to `backend.(f)counter_names` are, after all, basically just asking for the same constant list over and over again.

```
       before           after         ratio
     [f602e959]       [c0ae74fc]
     <nawtrey_issue_440_get_generic_record_perf>       <treddy_no_api_change_issue_440>
-     1.02±0.01ms         899±20μs     0.88  report.ModReadAllRecords.time_mod_read_all_records('dict', 'tests/input/sample-dxt-simple.darshan', 'MPI-IO', True)
-     1.26±0.01ms         1.10±0ms     0.87  report.ModReadAllRecords.time_mod_read_all_records('pandas', 'examples/example-logs/ior_hdf5_example.darshan', 'POSIX', False)
-         347±3μs          301±1μs     0.87  report.ModReadAllRecords.time_mod_read_all_records('numpy', 'examples/example-logs/dxt.darshan', 'STDIO', True)
-         276±8μs          240±4μs     0.87  report.ModReadAllRecords.time_mod_read_all_records('numpy', 'examples/example-logs/ior_hdf5_example.darshan', 'STDIO', False)
-         351±1μs          303±3μs     0.87  report.ModReadAllRecords.time_mod_read_all_records('numpy', 'examples/example-logs/ior_hdf5_example.darshan', 'STDIO', True)
-         286±5μs          245±3μs     0.86  report.ModReadAllRecords.time_mod_read_all_records('dict', 'examples/example-logs/ior_hdf5_example.darshan', 'STDIO', False)
-         289±2μs          246±1μs     0.85  report.ModReadAllRecords.time_mod_read_all_records('dict', 'examples/example-logs/dxt.darshan', 'STDIO', False)
-         362±9μs          305±3μs     0.84  report.ModReadAllRecords.time_mod_read_all_records('dict', 'examples/example-logs/ior_hdf5_example.darshan', 'STDIO', True)
-        404±20μs          328±3μs     0.81  report.ModReadAllRecords.time_mod_read_all_records('dict', 'examples/example-logs/ior_hdf5_example.darshan', 'MPI-IO', True)
-        405±10μs          316±6μs     0.78  report.ModReadAllRecords.time_mod_read_all_records('numpy', 'examples/example-logs/ior_hdf5_example.darshan', 'MPI-IO', True)
-        305±30μs          235±7μs     0.77  report.ModReadAllRecords.time_mod_read_all_records('numpy', 'examples/example-logs/dxt.darshan', 'STDIO', False)
-         418±4μs          318±3μs     0.76  report.ModReadAllRecords.time_mod_read_all_records('dict', 'examples/example-logs/dxt.darshan', 'MPI-IO', True)
-         345±3μs          260±7μs     0.75  report.ModReadAllRecords.time_mod_read_all_records('dict', 'examples/example-logs/ior_hdf5_example.darshan', 'MPI-IO', False)
-         332±4μs          249±7μs     0.75  report.ModReadAllRecords.time_mod_read_all_records('numpy', 'examples/example-logs/dxt.darshan', 'MPI-IO', False)
-         412±5μs          307±8μs     0.75  report.ModReadAllRecords.time_mod_read_all_records('numpy', 'examples/example-logs/dxt.darshan', 'MPI-IO', True)
-        447±10μs         331±10μs     0.74  report.ModReadAllRecords.time_mod_read_all_records('dict', 'examples/example-logs/dxt.darshan', 'POSIX', True)
-         343±2μs          250±1μs     0.73  report.ModReadAllRecords.time_mod_read_all_records('numpy', 'examples/example-logs/ior_hdf5_example.darshan', 'POSIX', False)
-         444±7μs          322±5μs     0.73  report.ModReadAllRecords.time_mod_read_all_records('dict', 'examples/example-logs/ior_hdf5_example.darshan', 'POSIX', True)
-         348±4μs          252±2μs     0.72  report.ModReadAllRecords.time_mod_read_all_records('dict', 'examples/example-logs/dxt.darshan', 'MPI-IO', False)
-         335±8μs          241±3μs     0.72  report.ModReadAllRecords.time_mod_read_all_records('numpy', 'examples/example-logs/ior_hdf5_example.darshan', 'MPI-IO', False)
-         426±9μs          305±2μs     0.72  report.ModReadAllRecords.time_mod_read_all_records('numpy', 'examples/example-logs/dxt.darshan', 'POSIX', True)
-        368±10μs          259±4μs     0.70  report.ModReadAllRecords.time_mod_read_all_records('dict', 'examples/example-logs/ior_hdf5_example.darshan', 'POSIX', False)
-         366±8μs          251±6μs     0.69  report.ModReadAllRecords.time_mod_read_all_records('numpy', 'examples/example-logs/dxt.darshan', 'POSIX', False)
-         373±8μs         253±30μs     0.68  report.ModReadAllRecords.time_mod_read_all_records('dict', 'examples/example-logs/dxt.darshan', 'POSIX', False)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

```diff
--- a/darshan-util/pydarshan/darshan/backend/cffi_backend.py
+++ b/darshan-util/pydarshan/darshan/backend/cffi_backend.py
@@ -4,6 +4,8 @@ using the functions defined in libdarshan-util.so
 and is interfaced via the python CFFI module.
 """
 
+import functools
+
 import cffi
 import ctypes
 
@@ -300,15 +302,13 @@ def log_get_record(log, mod, dtype='numpy'):
 
 
 
-def log_get_generic_record(log, mod_name, c_cols=None, fc_cols=None, dtype='numpy'):
+def log_get_generic_record(log, mod_name, dtype='numpy'):
     """
     Returns a dictionary holding a generic darshan log record.
 
     Args:
         log: Handle returned by darshan.open
         mod_name (str): Name of the Darshan module
-        c_cols: list of counter keys/column names
-        fc_cols: list of fcounter keys/column names
 
     Return:
         dict: generic log record
@@ -347,10 +347,8 @@ def log_get_generic_record(log, mod_name, c_cols=None, fc_cols=None, dtype='nump
     for i in range(flst.size):
         flst[i] = rbuf[0].fcounters[i]
 
-    if c_cols is None:
-        c_cols = counter_names(mod_name)
-    if fc_cols is None:
-        fc_cols = fcounter_names(mod_name)
+    c_cols = counter_names(mod_name)
+    fc_cols = fcounter_names(mod_name)
 
     if dtype == "numpy":
         rec['counters'] = clst
@@ -381,6 +379,7 @@ def log_get_generic_record(log, mod_name, c_cols=None, fc_cols=None, dtype='nump
     return rec
 
 
+@functools.lru_cache(maxsize=32)
 def counter_names(mod_name, fcnts=False, special=''):
     """
     Returns a list of available counter names for the module.
@@ -426,6 +425,7 @@ def counter_names(mod_name, fcnts=False, special=''):
     return names
 
 
+@functools.lru_cache(maxsize=32)
 def fcounter_names(mod_name):
     """
     Returns a list of available floating point counter names for the module.
diff --git a/darshan-util/pydarshan/darshan/report.py b/darshan-util/pydarshan/darshan/report.py
index 5f08674..69c5e00 100644
--- a/darshan-util/pydarshan/darshan/report.py
+++ b/darshan-util/pydarshan/darshan/report.py
@@ -604,13 +604,13 @@ class DarshanReport(object):
 
 
         # fetch records
-        rec = backend.log_get_generic_record(self.log, mod, c_cols=cn, fc_cols=fcn, dtype=dtype)
+        rec = backend.log_get_generic_record(self.log, mod, dtype=dtype)
         while rec != None:
             self.records[mod].append(rec)
             self._modules[mod]['num_records'] += 1
 
             # fetch next
-            rec = backend.log_get_generic_record(self.log, mod, c_cols=cn, fc_cols=fcn, dtype=dtype)
+            rec = backend.log_get_generic_record(self.log, mod, dtype=dtype)
 
 
         if self.lookup_name_records:
```

Reply:
It didn't occur to me to cache the counter/fcounter names, but this is safer (and faster) than what I had previously suggested. Thanks for the help here!

Reply:
I'll resolve this since these changes have been implemented and most are just reverting back to the original code.